codec-http2/src/test/java/io/netty/handler/codec/http2/Http2MultiplexCodecTest.java

```diff
@@ -583,6 +584,57 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
         assertFalse(channelActive.get());
     }
 
+    @Test
+    public void channelInactiveHappensAfterExceptionCaughtEvents() throws Exception {
+        final AtomicInteger count = new AtomicInteger(0);
+        final AtomicInteger exceptionCaught = new AtomicInteger(-1);
+        final AtomicInteger channelInactive = new AtomicInteger(-1);
+        final AtomicInteger channelUnregistered = new AtomicInteger(-1);
+
+        LastInboundHandler inboundHandler = streamActiveAndWriteHeaders(inboundStream);
+        Http2StreamChannel childChannel = (Http2StreamChannel) inboundHandler.channel();
+
+        // Add a handler that marks when events happen
+        childChannel.pipeline().addFirst("sniffer", new ChannelInboundHandlerAdapter() {
+
+            @Override
+            public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+                channelInactive.set(count.getAndIncrement());
+                super.channelInactive(ctx);
+            }
+
+            @Override
+            public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                exceptionCaught.set(count.getAndIncrement());
+                super.exceptionCaught(ctx, cause);
+            }
+
+            @Override
+            public void channelUnregistered(ChannelHandlerContext ctx) throws Exception {
+                channelUnregistered.set(count.getAndIncrement());
+                super.channelUnregistered(ctx);
+            }
+        });
+
+        childChannel.pipeline().addBefore("sniffer", "thrower", new ChannelInboundHandlerAdapter() {
```

Comment:
You can simplify this by just adding this handler first.

Reply:
How about simple `addFirst`?

Reply:
Sure... that said I think just adding it first is simpler. But whatever you prefer

Reply:
I'm not 100% sure what you mean, I need the pipeline to be in this order (and note that `streamActiveAndWriteHeaders` adds a `LastInboundHandler` to the pipeline) so adding the 'thrower' first would necessitate using `addAfter` to place the 'sniffer' handler.

Reply:
Basically just add the handlers in the opposite order and use `addLast`:


```java
childChannel.pipeline().addLast(new ChannelInboundHandlerAdapter() { }); // this is the thrower
childChannel.pipeline(). addLast(new ChannelInboundHandlerAdapter() { }); // this is the sniffer

```

Reply:
As it was written that wouldn't work because there is already a `ListInboundHandler` (added by the `streamActiveAndWrite that would soak up events. I've changed it to just an outbound stream (via `newOutboundStream`) and we can do what you propose which is cleaner. 

Thanks for the feedback!